lib/bap_image/llvm_binary_stubs.hpp

```diff
@@ -0,0 +1,449 @@
+    const pe32_header* hdr = 0;
+    if (error_code ec = obj -> getPE32Header(hdr))
+        llvm_binary_fail(ec);
+    if (!hdr)
+        llvm_binary_fail("PE header not found");
+    return hdr->AddressOfEntryPoint;
+};
+
+template <typename T>
+struct image : image_base {
+    explicit image(const T* obj) : obj_(obj) {}
+    Triple::ArchType arch() const {
+        return image_arch(obj_);
+    }
+
+    uint64_t entry() const {
+        return image_entry(obj_);
+    }
+
+    std::vector<seg::segment> segments() const {
+        return seg::read(obj_);
+    }
+
+    std::vector<sym::symbol> symbols() const {
+        return sym::read(obj_);
+    }
+
+    std::vector<sec::section> sections() const {
+        return sec::read(obj_);
+    }
+protected:
+    const T *obj_;
+};
+
+template <typename T>
+std::shared_ptr<image_base> create_image(const ObjectFile* obj) {
+    if (const T* ptr = dyn_cast<T>(obj))
+        return std::make_shared< image<T> >(ptr);
+    llvm_binary_fail("Unrecognized object format");
+}
+
+std::shared_ptr<image_base> create_image_elf(const ObjectFile* obj) {
+    if (const ELF32LEObjectFile *elf = dyn_cast<ELF32LEObjectFile>(obj))
+        return create_image<ELF32LEObjectFile>(elf);
+
+    if (const ELF32BEObjectFile *elf = dyn_cast<ELF32BEObjectFile>(obj))
+        return create_image<ELF32BEObjectFile>(elf);
+
+    if (const ELF64LEObjectFile *elf = dyn_cast<ELF64LEObjectFile>(obj))
+        return create_image<ELF64LEObjectFile>(elf);
+
+    if (const ELF64BEObjectFile *elf = dyn_cast<ELF64BEObjectFile>(obj))
+        return create_image<ELF64BEObjectFile>(elf);
+    llvm_binary_fail("Unrecognized ELF format");
+}
+
+std::shared_ptr<image_base> create_image(const ObjectFile* obj) {
+    if (obj->isCOFF())
+        return create_image<COFFObjectFile>(obj);
+    if (obj->isELF())
+        return create_image_elf(obj);
+    if (obj->isMachO())
+        return create_image<MachOObjectFile>(obj);
+    llvm_binary_fail("Unrecognized object format");            
+}
+
+std::shared_ptr<image_base> create_image(const Archive* arch) {
+    llvm_binary_fail("Archive loading unimplemented");
+}
+
+std::shared_ptr<image_base> create_image(const Binary* binary) {
+    if (const Archive *arch = dyn_cast<Archive>(binary))
+        return create_image(arch);
+    if (const ObjectFile *obj = dyn_cast<ObjectFile>(binary))
+        return create_image(obj);
+    llvm_binary_fail("Unrecognized binary format");
+}
+} //namespace img
+
+
+namespace binary {
+using namespace llvm;
+using namespace llvm::object;
+
+Binary* llvm_binary_create(const char* data, std::size_t size) {
+    StringRef data_ref(data, size);
+    MemoryBuffer* buff(MemoryBuffer::getMemBufferCopy(data_ref, "binary"));
+    OwningPtr<object::Binary> binary;
+    if (error_code ec = createBinary(buff, binary))
+        llvm_binary_fail(ec);
+    return binary.take();
+}
+
+Triple::ArchType llvm_binary_arch(Binary* binary) {
+    return img::create_image(binary)->arch();
```

Comment:
is it good to create image wrapper in each accessor function? Why don't you call this in `llvm_binary_create`?